wicket/src/runner.rs

```diff
@@ -0,0 +1,253 @@
+                                    .update_inventory(inventory)
+                                {
+                                    error!(
+                                        self.log,
+                                        "Failed to update inventory: {e}",
+                                    );
+                                } else {
+                                    redraw = true;
+                                }
+                            };
+
+                            self.state
+                                .service_status
+                                .reset_wicketd(wicketd_received.elapsed());
+                            self.state
+                                .service_status
+                                .reset_mgs(mgs_received.elapsed());
+                        }
+                        InventoryEvent::Unavailable { wicketd_received } => {
+                            self.state
+                                .service_status
+                                .reset_wicketd(wicketd_received.elapsed());
+                        }
+                    }
+
+                    redraw |= self.state.service_status.should_redraw();
+
+                    if redraw {
+                        // Inventory or status bar changed. Redraw the screen.
+                        self.screen.draw(&self.state, &mut self.terminal)?;
+                    }
+                }
+                _ => info!(self.log, "{:?}", event),
+            }
+        }
+        Ok(())
+    }
+
+    fn handle_action(&mut self, action: Option<Action>) -> anyhow::Result<()> {
+        let Some(action) = action else {
+         return Ok(());
+        };
+
+        match action {
+            Action::Redraw => {
+                self.screen.draw(&self.state, &mut self.terminal)?;
+            }
+            Action::Update(_component_id) => todo!(),
+        }
+        Ok(())
+    }
+
+    fn start_tokio_runtime(&mut self) {
+        let events_tx = self.events_tx.clone();
+        let log = self.log.clone();
+        let wicketd_manager = self.wicketd_manager.take().unwrap();
+        self.tokio_rt.block_on(async {
+            run_event_listener(log.clone(), events_tx).await;
+            tokio::spawn(async move {
+                wicketd_manager.run().await;
+            });
+        });
+    }
+}
+
+fn is_control_c(key_event: &KeyEvent) -> bool {
+    key_event.code == KeyCode::Char('c')
+        && key_event.modifiers == KeyModifiers::CONTROL
+}
+
+/// Listen for terminal related events
+async fn run_event_listener(log: slog::Logger, events_tx: Sender<Event>) {
+    info!(log, "Starting event listener");
+    tokio::spawn(async move {
+        let mut events = EventStream::new();
+        let mut ticker = interval(Duration::from_millis(30));
+        loop {
+            tokio::select! {
+                _ = ticker.tick() => {
+                        if events_tx.send(Event::Tick).is_err() {
```

Comment:
Indentation is off here and in the arm below; I think rustfmt gives up on `select!` bodies some (or all?) of the time.

Reply:
Yeah, rustfmt doesn't work inside macros. Fixed in [fd4ea2d](https://github.com/oxidecomputer/omicron/pull/2440/commits/fd4ea2d1c6e5f6a298b987e48eeefb5ee6bbeaf0)